tests/unit/test_ast.py

```diff
@@ -93,13 +93,26 @@ def test_match_bad_arg(self):
         assert "must take a regex" in info["errors"][0]
 
     def test_contains_bad(self):
-        a = ast.ContainsOperator(ast.Literal("foo"), ast.Empty())
+        a = ast.ContainsOperator("contains", ast.Literal("foo"), ast.Empty())
         valid, info = a.validate()
         assert not valid
         assert "Contains operator must take" in info["errors"][0]
 
     def test_contains_valid_args(self):
-        a = ast.ContainsOperator(ast.Literal("foo"), ast.Literal("bar"))
+        a = ast.ContainsOperator("contains", ast.Literal("foo"), ast.Literal("bar"))
```

Comment:
I'm not sure I understand why this is valid. IIUC, this is because `pypred` allows you to use contains to compare strings litterals?

Reply:
This is valid because `Literal` nodes are used for literal string values (`'foo'`) and for string that are used as an identifier and resolve to another value (`some_slot`). For example: `some_slot is 'foo'`

Reply:
I decided to not handle the case for `contains anyof` and `contains allof` to reduce complexity, so we couldn't do something like `{'foo' 'bar'} contains allof some_list_slot`

Reply:
I admit that this is not great, in my opinion nodes that are identifiers should be different than nodes for string literals, but that's how it's handled currently in the library 😕

Reply:
I see. This seems fine temporarily, but it's not intuitive. I expect to be able to use identifiers and literals interchangeably everywhere. I suggest we fix that in the next story. I do agree with you, the best solution for this would be to deal with identifiers in the same way everywhere, having a node type for it makes it much clearer/easier to deal with.

Reply:
Sorry another question to clarify : Can we do `some_set_slot contains anyof some_other_set_slot`?

Reply:
If I understand [this](https://github.com/dialoguemd/pypred/pull/2/files#diff-9f0c6436c1f69ab65cf72a07463aee64c776009f8f9a797d74590b3ebf577c8eR107) correctly, I think no?

Reply:
For the moment no, it's not allowed, but will be in next story while doing 👇  

> I expect to be able to use identifiers and literals interchangeably everywhere. I suggest we fix that in the next story.